src/libraries/System.Net.Quic/src/System/Net/Quic/QuicConnection.cs

```diff
@@ -636,6 +643,28 @@ private static unsafe int NativeCallback(QUIC_HANDLE* connection, void* context,
         }
     }
 
+    private void CompleteAcceptQueue(Exception? ex, bool drain)
+    {
+        _acceptQueue.Writer.TryComplete(ex);
+
+        if (drain)
+        {
+            // This should be only called after connection SHUTDOWN_COMPLETE has been indicated.
+            // At that point, all streams have been already shut down internally and we need
+            // only to close the handle via dispose, so DisposeAsync below should complete
+            // synchronously (which is necessary for this method to be callable from MsQuic
+            // event callback).
+            while (_acceptQueue.Reader.TryRead(out QuicStream? stream))
+            {
+                ValueTask task = stream.DisposeAsync();
+                Debug.Assert(task.IsCompletedSuccessfully);
+                task.GetAwaiter().GetResult();
+            }
```

Comment:
I am not 100% happy with this, having to wait for `SHUTDOWN_COMPLETE` to be able to dispose synchronously may delay callers stuck on 

https://github.com/dotnet/runtime/blob/79a715666e5b8d0267f72cde834492487d3a6f0e/src/libraries/System.Net.Quic/src/System/Net/Quic/QuicConnection.cs#L435

The delay between `SHUTDOWN_INITIATED_BY_*` and `SHUTDOWN_COMPLETE` depends on the RTT time between peers and can be up to 15ms (per MsQuic code).

However, given that the situation is a bit rare, I would not bend backwards to fix it.

Reply:
You need to do this only because we're abusing `_acceptQueue` to propagate shutdown exception to the caller of `OpenStream`. Wouldn't it be better to just store the exception somewhere else? E.g. additional `Exception _shutdownException`.

---
My musings on this whole shutdown ordeal, feel free to ignore.
We also have shutdown cancellation token, where we also do some hacky things to propagate exception from shutdown: 
https://github.com/dotnet/runtime/blob/e58be4792208e1cda48620901e6c7822262809e5/src/libraries/System.Net.Quic/src/System/Net/Quic/QuicListener.cs#L256-L262
So maybe, we could somehow solve this all together somehow...

Reply:
Also found this related API suggestion: https://github.com/dotnet/runtime/issues/95263